feat(dataset): dataset creation with s3 storage backend

mohammad-sdsc opened the pull request:
# Description

Allows creating datasets with S3 storage backends. Currently, only supports `s3://<bucket-name>/<path>` URI format. Uses `RClone` to interact with S3.

### NOTE:
`RClone` cannot check if a directory exists for storage backends that don't support empty directories like S3. So, it's not possible to fail if the directory doesn't exists as was required in the story.

Fixes #2970 
Fixes #2972 

### TODO
Test with private S3 buckets and valid S3 credentials.

Panaetius commented on renku/infrastructure/storage/base.py:
```diff
@@ -0,0 +1,96 @@
+            name = f"-{key}" if len(key) == 1 else f"--{key.replace('_', '-')}"
+            return [name] if value is True else [name, f"{value}"]
+
+    args = [transform_kwarg(k, v) for k, v in kwargs.items()]
+    return [a for arg in args for a in arg]
```
```suggestion
    args =[]
    for k, v in kwargs.items():
        args.extend(transform_kwarg(k, v))
    return args
```

not that it matters at all, feel free to ignore, i just find it slightly more readable than the list comprehension in this case.

Panaetius commented on renku/core/dataset/providers/api.py:
```diff
@@ -215,3 +234,54 @@ def get_access_token_url(self) -> str:
+class ProviderCredentials(abc.ABC, UserDict):
+    """Credentials of a provider.
+
+    NOTE: A ``None`` value for a key means that it is not set; however, an empty value, "", means that the key was set.
```
Together with the other comment on `None` instead of `""`. We do have `renku.core.util.util.NO_VALUE` (And `NoValueType`) for this purpose, which allows showing intent more clearly and also reflects this in the type signature.

mohammad-sdsc replied:
`""` is a valid value. So, I've changed `None` to `NO_VALUE`.

Panaetius commented on renku/core/dataset/providers/api.py:
```diff
@@ -215,3 +234,54 @@ def get_access_token_url(self) -> str:
+
+    @staticmethod
+    @abc.abstractmethod
+    def get_credentials_names() -> Tuple[str, ...]:
```
Should this just return `List[str]` as per the docstring?

mohammad-sdsc replied:
This won't change and makes sense to be a tuple. I've updated the docstring.

Panaetius reviewed (changes requested):
Looks really nice and should be easy to extend, thank you. Only some minor comments.

olevski commented on renku/infrastructure/storage/base.py:
```diff
@@ -0,0 +1,96 @@
+
+
+class BaseStorage(IStorage):
+    """Base external storage handler class."""
```
This currently has a lot of rclone-specific things. Should it then be called `RcloneBaseStorage` or something like that. Or if this is supposed to be more general then the rclone-specific stuff should then move someplace else?

mohammad-sdsc commented on renku/core/dataset/providers/s3.py:
```diff
@@ -0,0 +1,96 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright 2020 - Swiss Data Science Center (SDSC)
```
I've changed all of them to `2017-2022`

olevski commented on renku/domain_model/dataset.py:
```diff
@@ -318,6 +318,7 @@ class Dataset(Persistent):
     """Represent a dataset."""
 
     date_modified: Optional[datetime] = None  # type: ignore
+    storage: Optional[str] = None
```
Do we really need to have `storage` as both a class and instance variable? Isn't it simpler to just have the instance variable?

But maybe I am missing something. This also seems to happen with a few other class/instance variables on dataset.

mohammad-sdsc replied:
We need the class variable for backward compatibility: When we load an object from an older project that doesn't have the `storage` property, the class variable makes sure that this property is set to `None`. The library that we use to store Renku metadata (i.e. ZODB/zope/persistent) doesn't call the `__init__` method when deserializing an object (it creates an empty object and calls `__setstate__` upon the first access to the object). So, we need the class variable for every new attribute that we add to our models.

olevski commented:
Also (if I am not mistaken) the credentials can only be added for a single s3 bucket. So I believe that the way this is setup is that I can only add buckets (with credentials) from a single region and provider.

If I am wrong about this lets touch base over slack @mohammad-sdsc because I probably misunderstood of how the credential provider is supposed to be used.

But my impression is that `s3://` as a prefix is only tied to one set of credentials. And we are allowing only `s3://` to be used as a prefix for s3 buckets.

mohammad-sdsc commented:
> Also (if I am not mistaken) the credentials can only be added for a single s3 bucket. So I believe that the way this is setup is that I can only add buckets (with credentials) from a single region and provider.
> 
> If I am wrong about this lets touch base over slack @mohammad-sdsc because I probably misunderstood of how the credential provider is supposed to be used.
> 
> But my impression is that `s3://` as a prefix is only tied to one set of credentials. And we are allowing only `s3://` to be used as a prefix for s3 buckets.

That's true, atm we only can have one set of credentials for S3. We can have multiple credentials by overriding `ProviderCredentials::get_credentials_section_name` for S3 (the comment there reads: `NOTE: This methods should be overridden by subclasses to allow multiple credentials per providers if needed.`). I added bucket name as the part of key before but then we needed to enter the credentials for each bucket so I removed it. I can easily put it back or implement any other mechanism.

olevski commented:
This LGTM but if you want to wait also for Ralf's approval that is def fine by me.
